Pull nexus library into mantid

peterfpeterson opened the pull request:
This pulls in the minimum amount of code from https://github.com/nexusformat/code/ to replace the nexus dependency in mantid. The criteria for success is that the nexus library is no longer a dependency and the source is changed enough to compile without warnings, pass cppcheck, and all tests pass. There is a lot of work left to do to make the code performant and bug free.

Refs #38332 and [EWM8446](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=8446)

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work

Random comment: Changing `.c` to `.cpp` put cmake into a mode where it now acknowledges that these are c++ files and more changes had to be made.

Much of the information on how to build nexus was taken from the [nexus conda package](https://github.com/conda-forge/nexus-feedstock/blob/main/recipe/meta.yaml) which is [v4.4.3 of nexus](https://github.com/nexusformat/code/releases/tag/v4.4.3) (Aug 15, 2016) despite there being changes as recent as Jan 26, 2020. The code from the most recent changes, rather than the tag, was the starting point.

<!-- Please provide a more detailed description of the work that has been undertaken.
-->

### To test:

This build will check that system tests and packaging work on all OS:
https://builds.mantidproject.org/job/build_packages_from_branch/961/


<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

peterfpeterson committed:
Update developer dependencies

peterfpeterson committed:
Add conda-verify to quiet packaging warning

peterfpeterson committed:
Replace nexus with its dependencies

Refs #38332

The dependencies of https://github.com/conda-forge/nexus-feedstock/blob/main/recipe/meta.yaml
* hdf4 - added
* hdf5 - already present
* libjpeg-turbo - required by hdf4
* zlib - added
* libxml2 - required by libxslt
* ncurses - required by readline
* readline - required by python

peterfpeterson committed:
First copying of files into NexusCpp subpackage

This creates a shared object with a bunch of compiler warnings and
doesn't bother with getting tests in.

martyngigg commented on .gitignore:
```diff
@@ -208,6 +208,9 @@ rel/
 miniconda3/
 
+# in source build of conda pacakges
+conda/recipes/
```
Won't this ignore all changes to the sources in `conda/recipes` too?

peterfpeterson replied:
Things in `.gitignore` are default ignored. You can still add them explicitly.

peterfpeterson committed:
Add tests which compile but fail

peterfpeterson committed:
Remove files that implement xml backend

peterfpeterson committed:
Remove ifdef for xml

peterfpeterson committed:
Add test for unlimited arrays

peterfpeterson committed:
Fix cppcheck warnings

peterfpeterson committed:
Remove NeXusStream class

peterfpeterson committed:
Fix compiler warnings and cppcheck

peterfpeterson committed:
Give cpp tests standard extension

peterfpeterson committed:
Move contents of nxconfig into napiconfig

peterfpeterson committed:
Bare minimum to get the code to compile and link as c++

peterfpeterson committed:
Lots of little changes to make it more c++ ish

peterfpeterson committed:
Make test errors clearer

peterfpeterson committed:
Change import from nexus to MantidNexusCpp

peterfpeterson committed:
Fix cppcheck issues

peterfpeterson committed:
Fix cppcheck issues in nexuscpp

peterfpeterson committed:
Undo cppcheck fix

peterfpeterson committed:
Refactor printing into utility function

peterfpeterson committed:
Create external data files on the fly

Rather than storing them in the testing data repository, create them and
delete on completion of the test

peterfpeterson committed:
Always return that creating the attribute was ok

peterfpeterson committed:
Fix cppcheck issues

peterfpeterson committed:
Remove incorrect imports

peterfpeterson committed:
Fix compiler warnings

peterfpeterson committed:
Stop using deprecated functions

peterfpeterson committed:
Fix cppcheck issues

peterfpeterson committed:
Remove system nexus finder

peterfpeterson committed:
Change error reporting function to take const string

peterfpeterson committed:
Remove deprecated function

peterfpeterson committed:
Make final parameter const for clang

peterfpeterson committed:
Do not capture exception that is rethrown

peterfpeterson committed:
Import nxconfig to help windows build

peterfpeterson committed:
Change order of modules

peterfpeterson committed:
Skip clang compiler rather than intel

peterfpeterson committed:
Copy configuration from Kernel to fix windows

peterfpeterson committed:
Remove wrong header file

peterfpeterson committed:
Change signature to quiet msvs warnings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

peterfpeterson committed:
Remove unused setnumberformat function

peterfpeterson committed:
Use strdup variant on windows

peterfpeterson committed:
Fix compiler warnings on windows

peterfpeterson committed:
Change default create to hdf5

peterfpeterson committed:
Skipping external file tests on windows

peterfpeterson committed:
Try to make windows test errors clearer

peterfpeterson committed:
No longer configure packagesetup.py because nexus is gone

peterfpeterson committed:
Turn off the c-hdf5 test on windows

peterfpeterson committed:
Update cppcheck suppression

thomashampson reviewed (changes requested):
Some minor things to address. Also we should create a maintenance issue to re-enable the tests that are skipped on Windows.

thomashampson commented:
Checking that the system tests and packaging all work here:
https://builds.mantidproject.org/job/build_packages_from_branch/961/

peterfpeterson commented on Framework/NexusCpp/src/napi5.cpp:
```diff
@@ -0,0 +1,2587 @@
+  /* Close attribute  */
+  iRet += H5Aclose(attr1);
+  killAttVID(pFile, vid);
+  // TODO QUESTION always return that it is ok?
```
I only know that returning the actual value makes several tests unreliable. Since we're replacing much of `napi5.cpp` I'm not too worried about it yet.

peterfpeterson committed:
Remove commented out code and change some magic numbers to constants

thomashampson reviewed (changes requested):
It looks like the Windows packaging has failed for some reason:
https://builds.mantidproject.org/job/build_packages_from_branch/961/
I'll look into that

thomashampson reviewed (approved):
All system tests passed in the linked build, and the conda packaging was fine on all except the Windows workbench stage. It failed because the commit was no longer available when it reached that stage because of a force push that overwrote the head commit. There is no need to re-run the job, I think it has done enough to demonstrate that the code is working fine.
